### ISSUE TITLE
More terse function declarations.

### ISSUE THREAD

**Issue by mckoss** [firebase/bolt#40]
Given that Bolt is a non-procedural language, the current procedural looking functions are not ideal.

Note that ES6 defines a more concise form of function declaration, [arrow functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions):

```
function f(x) { return exp; }  // function
```
is the same as
```
f = (x) => { return exp; } // function
```
or even more concisely
```
f = (x) => exp;
```

I would personally prefer the even more generic looking:
```
f(x) = exp;
```

Methods, though, still have the un-needed { return ... }:
```
read() { return exp; }
```
The logical extension (not in ES6, AFIK) is to allow method definitions like:
```
read() = exp;
```
or, for zero-argument functions:
```
read = exp;
```

**Comment by mckoss** [comment]
@tomlarkworthy @matjaz @TristonianJones Thoughts?

**Comment by matjaz** [comment]
In ES6 you can do
`f = x => exp`
or with zero arguments
`f = () => exp`
Arrow funcs. also have automatic context binding (this), but since bolt doesn't have a scope, this is not important.

I like `read() = exp;` and also `read = exp;` and since bolt doesn't have state, the only assignable thing is function.

Similar could be done with type/path methods.

**Comment by tomlarkworthy** [comment]
For blaze I went for 
```f(x) : exp```

because := is a notation is some circles, and : part of YAML. I would have avoided f(x) = exp as I think = is a bit overloaded. If you replaced it with => it would look like:

```
f(x) => x + 1
```

which I think its pretty good but its not valid typescript. It is relatively easy to implement though.

If we wanted valid typescript I would go for Matjaz suggestion of:

```
f = (x: Number, y: String) => x + y
``` 

Although that above example is really doing assignment (f *=* ) of a function, so the language would really have to have the notion of a function as a first class type, i.e. lambda calculus. I am super keen on lambda calculus in the language, but it's work. I think it would allow some nice abstractions to be built by users. Doing some hacky shortcuts to make it look like lambda calculus but without actually supporting function types I think would cause problems down the road.

To summarize, I think 
do `f(x) => x + 1` if you don't want to implement lambda calculus
do `f = (x: Number, y: String) => x + y` if you implement all the glory of lambda calculus (its not particularly hard to do, but probably needs a revision of the internal representation). 

Tom

**Comment by mckoss** [comment]
Functions are already first class in bolt:

```
function f(g, x) { return g(x)*3; }
function y(x) { return x + 2; }

path / { validate() { return f(y,this); }}
```
generates:
```
{
  "rules": {
    ".validate": "(newData.val() + 2) * 3"
  }
}
```
We don't allow recursion is expression expansion, so I don't think you can write a Y-combinator ;-)

**Comment by tomlarkworthy** [comment]
Oh that's so sweet. + 1000. Hmmm, well then 
``` f = (x: Number, y: String) => x + y ```

is probably the most consistent with existing languages then, although it is a little clumsy compared to ```f(x) = exp```. That's like Scala. I just don't like = on its own given =, ==, and  === . If we wanted our own terser style I would be happier with f(x) := exp
